File: array/11/11_container_with_most_water.cpp
#include <vector>
using namespace std;

class Solution {
public:
  int maxArea(vector<int> &height) {
    int max_area = 0;
    int i = 0;
    int j = height.size() - 1;
    while (i < j) {
      int h = min(height[i], height[j]);
      max_area = max(max_area, h * (j - i));
      while (height[i] <= h && i < j)
        i++;
      while (height[j] <= h && i < j)
        j--;
    }
    return max_area;
  }
};

File: array/11/11_container_with_most_water.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_area = 0
        i = 0
        j = len(height) - 1
        while i < j:
            min_height = min(height[i], height[j])
            area = min_height * (j - i)
            if area > max_area:
                max_area = area
            if height[i] < height[j]:
                i += 1
            else:
                j -= 1
        return max_area


height = [1, 8, 6, 2, 5, 4, 8, 3, 7]
main = Solution().maxArea(height)
print(main)
